File: 61.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


# 1. Compute the size of the list
# 2. Take k modulo the size
# 3. Rotate the list by this amount
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next or k == 0:
            return head
        n = 0
        curr = head
        while curr:
            n += 1
            curr = curr.next
        k = k % n
        if k == 0:
            return head
        s = head
        f = head
        while f and f.next:
            if k <= 0:
                s = s.next
            f = f.next
            k -= 1
        f = s.next
        s.next = None
        new_head = f
        while f and f.next:
            f = f.next
        f.next = s
        return new_head

File: 61.cpp
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr)
    {
    }
    ListNode(int x) : val(x), next(nullptr)
    {
    }
    ListNode(int x, ListNode *next) : val(x), next(next)
    {
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
  public:
    ListNode *rotateRight(ListNode *head, int k)
    {
        // base case:
        if (!head || !head->next || k == 0)
            return head;
        // find the size
        int n = 0;
        for (auto curr = head; curr; curr = curr->next)
            n += 1;
        // update k
        k %= n;
        if (k == 0)
            return head;
        auto s = head;
        auto f = head;
        while (f && f->next)
        {
            if (k <= 0)
            {
                s = s->next;
            }
            f = f->next;
            k -= 1;
        }
        f = s->next;
        s->next = nullptr;
        auto new_head = f;
        for (; f && f->next; f = f->next)
            ;
        f->next = head;
        return new_head;
    }
};